File: 0726-number-of-atoms/0726-number-of-atoms.py
class Solution:
    def countOfAtoms(self, f: str) -> str:
        stack = [1]
        mp, i = defaultdict(int), len(f) - 1
        
        while i >= 0:
            temp = ''
            while i >= 0 and f[i].isdigit():
                temp = f[i] + temp
                i -= 1
            if temp:
                digit = int(temp)
            else:
                digit = 0
            if i >= 0 and f[i] == ')':
                i -= 1
                stack.append(stack[-1] * max(1, digit))
            elif i >= 0 and f[i] == '(':
                i -= 1
                stack.pop()
            else:
                curr = ''
                while i >= 0 and f[i].islower():
                    curr = f[i] + curr
                    i -= 1
                if i >= 0 and f[i].isupper():
                    curr = f[i] + curr
                    i -= 1
                mp[curr] += stack[-1] * max(1, digit)

        ans = ''

        for curr in sorted(mp.keys()):
            ans = ans + curr
            if mp[curr] != 1:
                ans += str(mp[curr])

        return an

File: 0726-number-of-atoms/0726-number-of-atoms.cpp
class Solution {
public:
    string countOfAtoms(string f) {
        stack<int> stk;
        stk.push(1);
        map <string, int> mp;
        int i = f.size() - 1;

        while (i >= 0) {
            string temp = "";
            while (i >= 0 && isdigit(f[i])) {
                temp = f[i] + temp;
                i--;
            }
            int digit = temp.empty() ? 0 : stoi(temp);
            if (i >= 0 && f[i] == ')') {
                i--;
                stk.push(stk.top() * max(1, digit));
            } else if (i >= 0 && f[i] == '(') {
                i--;
                stk.pop();
            } else {
                string curr = "";
                while (i >= 0 && islower(f[i])) {
                    curr = f[i] + curr;
                    i--;
                }
                if (i >= 0 && isupper(f[i])) {
                    curr = f[i] + curr;
                    i--;
                }
                mp[curr] += stk.top() * max(1, digit);
            }
        }

        string ans = "";
        
        for (auto p : mp) {
            ans += p.first;
            if (p.second != 1) {
                ans += to_string(p.second);
            }
        }

        return ans;
    }
};
